onnxruntime/core/providers/qnn/builder/qnn_node_group/reshape_gemm_fusion.cc

```diff
@@ -0,0 +1,208 @@
+  if (!shape_proto) return false;
+  const auto* dtype = DataTypeImpl::TensorTypeFromONNXEnum(shape_proto->data_type())->GetElementType();
+  TensorShape shape = onnxruntime::utils::GetTensorShapeFromTensorProto(*shape_proto);
+  Tensor tensor(dtype, shape, std::make_shared<CPUAllocator>());
+  if (onnxruntime::utils::TensorProtoToTensor(onnxruntime::Env::Default(), graph_viewer.ModelPath(), *shape_proto,
+                                              tensor) != Status::OK()) {
+    return false;
+  }
+
+  InlinedVector<int64_t> output_shape;
+  if (tensor.IsDataType<int64_t>()) {
+    gsl::span<const int64_t> tensor_elems = tensor.DataAsSpan<int64_t>();
+    output_shape.insert(output_shape.end(), tensor_elems.begin(), tensor_elems.end());
+  } else if (tensor.IsDataType<int32_t>()) {
+    gsl::span<const int32_t> tensor_elems = tensor.DataAsSpan<int32_t>();
+    for (int32_t elem : tensor_elems) {
+      output_shape.emplace_back(static_cast<int64_t>(elem));
+    }
+  }
+
+  return !input_shape.empty() && output_shape.size() == 2 && input_shape.back() == output_shape.back();
+}
+
+#define ValidateOnQnn(qnn_model_wrapper, reshape_node_unit, gemm_node_unit) \
+  CreateOrValidateOnQnn((qnn_model_wrapper), (reshape_node_unit), (gemm_node_unit), true)
+#define CreateOnQnn(qnn_model_wrapper, reshape_node_unit, gemm_node_unit) \
+  CreateOrValidateOnQnn((qnn_model_wrapper), (reshape_node_unit), (gemm_node_unit), false)
+static Status CreateOrValidateOnQnn(QnnModelWrapper& qnn_model_wrapper, const NodeUnit& reshape_node_unit,
+                                    const NodeUnit& gemm_node_unit, bool validate) {
+  assert(reshape_node_unit.OpType() == "Reshape" && gemm_node_unit.OpType() == "Gemm");
+  const auto& node_name = utils::GetNodeName(gemm_node_unit);
+  const NodeUnitIODef& input_def = reshape_node_unit.Inputs()[0];
+  const NodeUnitIODef& weight_def = gemm_node_unit.Inputs()[1];
+  const NodeUnitIODef* bias_def_ptr = nullptr;
+  bool has_bias = gemm_node_unit.Inputs().size() == 3;
+  if (has_bias) {
+    bias_def_ptr = &gemm_node_unit.Inputs()[2];
+  }
+  const NodeUnitIODef& output_def = gemm_node_unit.Outputs()[0];
+
+  QnnTensorWrapper input_tensor;
+  QnnTensorWrapper bias_tensor;
+  QnnTensorWrapper output_tensor;
+  ORT_RETURN_IF_ERROR(qnn_model_wrapper.MakeTensorWrapper(input_def, input_tensor));
+  std::vector<uint32_t> weight_shape;
+  std::vector<uint8_t> unpacked_tensor;
+  std::string weight_tensor_name = weight_def.node_arg.Name();
+  ORT_RETURN_IF_NOT(qnn_model_wrapper.GetOnnxShape(weight_def.node_arg, weight_shape), "Failed to get weight shape");
+  Qnn_TensorType_t tensor_type = qnn_model_wrapper.GetTensorType(weight_tensor_name);
+  Qnn_DataType_t data_type = QNN_DATATYPE_FLOAT_32;
+  ORT_RETURN_IF_ERROR(utils::GetQnnDataType(false, weight_def.node_arg.TypeAsProto(), data_type));
+  const auto& weight_tensor_proto = qnn_model_wrapper.GetInitializerTensors().at(weight_tensor_name);
```

Comment:
Using GetConstantInitializer will fail one QNN test because some initializers in the test model are not constant. So I am using graph_viewer_.GetInitializedTensor instead. Not sure if this is also a problem for NNAPI or CoreML EP.

Reply:
@adrianlizarraga, I think Scott is right, we need to make sure the initializers in the test models are constant. One of the models failed to run is qnn_test_data\float32_models\facedetection_op8, which is in blob. I don't know how to update them, not sure if it's save to update the model directly, not sure if I have the permission to do that. Could you please advice how we can fix the test models?

Reply:
It's ".\RelWithDebInfo\onnx_test_runner -j 1 -v -e qnn -i "backend_path|D:\a\_work\_temp\qnn-v2.28.2.241116_win\lib\x86_64-windows-msvc\QnnCpu.dll" C:\data\float32_models" scanning the whole folders, including models and input data files, and there are also quantized models in another folder for another tests, quite a lot of files...